software_development/DEVELOPMENT.md

```diff
@@ -0,0 +1,325 @@
+
+### ServiceContext
+
+``ServiceContext`` provides access to detail on the current request, and the user session.
+
+When creating a ServiceContext you are responsible for manging its use on the current thread and any cleanup.
+
+Using auto-closable:
+```
+try(ServiceContext context = serviceMan.createServiceContext("md.thumbnail.upload", lang, request)){
+    ...
+}
+```
+
+Or manually:
+```
+ServiceContext context = serviceMan.createServiceContext("md.thumbnail.upload", lang, request);
+try {
+    context.setAsThreadLocal();
+    ...
+} finally {
+    context.clearAsThreadLocal();
+    context.clear();
+}
+```
+
+Many utility classes and methods expect a service context to be provide as a parameter, or to be available for the current thread.
+
+```
+context = ServiceContext.get();
+```
+
+ServiceContext is intended to be created using the current HTTPRequest:
+
+```
+try (ServiceContext context = serviceManager.createServiceContext("md.thumbnail.upload", lang, request)) {
+   DataManager dataMan = context.getBean(DataManager.class);
+   String version = dataMan.getVersion(id);
+   ...
+}
+```
+
+This pattern is so commonly used for API methods that ``ApiUtils`` provides a helper:
+```
+try (ServiceContext context = ApiUtils.createServiceContext(request)) {
+   // the context is alread set on the thread locale
+   ...
+}
+```
+
+When implementing event handlers, such as ``BeforeCommitTransactionListener`` there is no direct access to the current HTTPRequest. The ``createServiceContext( name, userId)`` method obtains the ``HTTPRequest`` from the ServletRequest holder if used during a request, or uses the provided  userId to look up user session when running in a background thread:
+
+```
+try (ServiceContext context = serviceManager.createServiceContext("approve_record", userId)) {
+   draftUtilities.replaceMetadataWithDraft(event.getMd());
+}
+```
+
+It can also be a challenge to provide a service context to a background job or activity. Use your exiesting service context to create a context for the background job or activity to use:
+
+```
+ServiceContext context = serviceManager.createServiceContext( "harvester."+type, serviceContext );
+harvester.initContext( context );
+```
+
+The background job or activity is responsible eventual cleanup:
+```
+public void shutdown() throws SchedulerException {
+   getScheduler().deleteJob(jobKey(getParams().getUuid(), HARVESTER_GROUP_NAME));
+   context.clear();
+   context = null;
+}
+```
+
+Here is the same approach using a Runnable:
+```
+final context = serviceManager.createServiceContext("later", serviceContext);
+return new Runnable(){
+    public void run() {
+        context.setAsThreadLocal();
+        try {
+            DataManager dataManager = context.getBean(DataManager.class);
+            ...
+        }
+        finally {
+          context.clearAsThreadLocal();
+          context.clear();
```

Comment:
Some questions:

1) Is it required to call `clearAsThreadLocal` only in case previously has been used `setAsThreadLocal` in the context, otherwise calling `clear` is enough?

2) When calling `clearAsThreadLocal` is also required to call `clear` after? Or there are cases where is not required?

Reply:
I will add more details.

1. Yes, but also some create methods like `ApiUtils`  have the side effect (documented) of calling `clearAsThreadLocal()`

2. Yes `clear` is required, it lifecycle end (the ooppose of `create`).

   Calling `clearAsThreadLocal()` is only required if `setAsThreadLocal()` has been used (directly or indirectly).